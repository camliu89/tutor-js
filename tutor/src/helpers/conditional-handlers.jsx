import React from 'react';
import invariant from 'invariant';
import { extend, last, first } from 'lodash';
import { Redirect } from 'react-router-dom';
import { asyncComponent } from './async-component';
import { CourseNotFoundWarning } from '../components/course-not-found-warning';
import Courses from '../models/courses-map';
import { OXMatchByRouter } from 'shared';

const StudentDashboard = asyncComponent(
  () => import('../screens/student-dashboard')
);

const getConditionalHandlers = (Router) => {
  const MatchForTutor = OXMatchByRouter(Router, null, 'TutorRouterMatch');

  const renderDashboard = (props) => {
    const { courseId } = props.params;

    extend(props, { courseId });
    const course = Courses.get(courseId);

    if (!course) {
      return <CourseNotFoundWarning />;
    }

    if (!props.match.isExact) {
      return (
        <MatchForTutor {...props} />
      );
    }

    if (course.isTeacher) {
      if (course.is_concept_coach) {
        return (
          <CCDashboard courseId={courseId} {...props} />
        );
      } else {
        return (
          <Redirect
            to={{
              pathname: Router.makePathname('viewTeacherDashboard', props.params),
              query: Router.currentQuery(),
            }} />
        );
      }
    } else {
      if (course.is_concept_coach) {
        return (
          <CCStudentRedirect courseId={courseId} />
        );
      } else {
        return <StudentDashboard {...props} />;
      }
    }
  };
  // eslint-disable-next-line react/prop-types
  const legacyReferenceBookRedirect = ({ params }) => {
    const parts = params.parts.split('/');
    const course = Courses.get(first(parts));
    invariant(course, `Did not find course for params '${params}'`);
    const chapterSection = parts.length > 1 ? last(parts) : 0;
    return (
      <Redirect
        to={{
          pathname: Router.makePathname(
            chapterSection ? 'viewReferenceBookSection' : 'viewReferenceBook',
            {
              ecosystemId: course.ecosystem_id,
              chapterSection,
            }),
          query: Router.currentQuery(),
        }}
      />
    );
  };

  return {
    dashboard() { return renderDashboard; },
    legacyReferenceBookRedirect() { return legacyReferenceBookRedirect; },
  };
};

export { getConditionalHandlers };
