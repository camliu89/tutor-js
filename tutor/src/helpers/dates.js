import moment from 'moment';
import 'twix';
import 'moment-timezone';
import { compact, reduce, min, max, map } from 'lodash';
import TimeHelper from './time';

export function getDay(oneMoment) {
  return moment(oneMoment)
    .startOf('day')
    .twix(moment(oneMoment).endOf('day'), { allDay: true });
}

export function getDurationFromMoments(listOfMoments) {
  return reduce(
    listOfMoments,
    (current, next) => current.union(getDay(next)),
    getDay(listOfMoments[0])
  );
}

export function findEarliest(dateThings) {
  return min(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}

export function findLatest(dateThings) {
  return max(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}


export function dateWithUnchangedTime(date, t) {
  if (moment.isMoment(date)) {
    date = date.format(TimeHelper.ISO_DATE_FORMAT);
  }
  return TimeHelper.getDateOnly(date);
}
