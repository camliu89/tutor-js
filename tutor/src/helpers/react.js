import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';
import { propTypes as mobxPropTypes } from 'mobx-react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import styled, { css } from 'styled-components';
import { idType } from 'shared';

export {
  React, cn, ReactDOM, mobxPropTypes, PropTypes,
  invariant, styled, css, idType,
};
export { observable, action, computed } from 'mobx';
export { observer, inject, Provider } from 'mobx-react';
