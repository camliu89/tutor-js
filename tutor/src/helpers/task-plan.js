import { isEmpty, get, pick, first, every, reduce, map } from 'lodash';
import moment from 'moment';
import Time from '../models/time';

export default {

  // returns the opens_at and due_at dates combined if they're identical,
  // separated by target_id if not
  dates(plan, options) {
    if (options == null) { options = {}; }
    if (isEmpty( get(plan, 'tasking_plans' ))) { return {}; }
    const pickFn = options.only ? tp => pick(tp, options.only) : tp => pick(tp, 'opens_at', 'due_at');

    const firstTasking = first( plan.tasking_plans );
    const isIdentical = every(plan.tasking_plans, function(tp) {
      if (options.only != null) {
        return tp[options.only] === firstTasking[options.only];
      } else {
        return (tp.opens_at === firstTasking.opens_at) && (tp.due_at === firstTasking.due_at);
      }
    });
    if (isIdentical) {
      return { all: pickFn(firstTasking) };
    } else {
      return reduce( plan.tasking_plans, function(memo, tp) {
        memo[tp.target_id] = pickFn(tp);
        return memo;
      }
      , {});
    }
  },

  calendarParams(course) {
    let date;
    const dueAtRange = course
      .teacherTaskPlans.array
      .map(plan => plan.dateRanges.due.start)
      .sort();

    if (dueAtRange.length) {
      date = dueAtRange[0];
    } else {
      date = Time.now;
    }
    return {
      to: 'calendarByDate',
      params: {
        courseId: course.id,
        date: moment(date).format('YYYY-MM-DD'),
      },
    };
  },

  earliestDueDate(plan) {
    const dates = map(get(plan, 'tasking_plans'), 'due_at');
    return first(dates.sort()) || '';
  },

};
