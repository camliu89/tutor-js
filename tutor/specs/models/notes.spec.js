import { Factory } from '../helpers';
import DATA from '../../api/notes.json';
import { keys } from 'lodash';
import Notes, { PageNotes, Note } from '../../src/models/notes';
import FeatureFlags from '../../src/models/feature_flags';
import Hypothesis from '../../src/models/notes/hypothesis';
jest.mock('../../src/models/feature_flags');
jest.mock('../../src/models/notes/hypothesis');

describe('Notes Model', () => {
  let notes;

  beforeEach(() => {
    FeatureFlags.is_highlighting_allowed = true;
    notes = new Notes({ course: Factory.course() });
  });

  it('does not request info when feature flag is off', () => {
    const fetch = jest.spyOn(PageNotes.prototype, 'fetch');
    FeatureFlags.is_highlighting_allowed = false;
    notes.forPageId(1);
    expect(fetch).not.toHaveBeenCalled();
    FeatureFlags.is_highlighting_allowed = true;
    notes.forPageId(2);
    expect(fetch).toHaveBeenCalled();
  });

  it('maps notes by section', () => {
    const page = notes.forPageId(1);
    page.set(1, Factory.note());
    page.set(2, Factory.note());
    page.set(3, Factory.note());
    expect(page.notesBySection).toMatchSnapshot();
  });

});
