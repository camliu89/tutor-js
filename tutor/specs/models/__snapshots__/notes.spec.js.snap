// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes Model maps notes by section 1`] = `
Object {
  "10.3": Array [
    Note {
      "chapter": undefined,
      "chapter_section": ChapterSection {
        "chapter": 10,
        "section": 3,
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ChapterSection@466.keys",
            "observers": Set {},
          },
          "name": "ChapterSection@466",
          "target": [Circular],
          "values": Map {
            "isNew" => false,
            "isEmpty" => false,
            "asString" => "10.3",
            "asArray" => Array [
              10,
              3,
            ],
          },
        },
      },
      "contents": Object {
        "content": "plants and animals appeared.",
        "endContainer": "./*[name()='p'][1]/text()[1]",
        "endOffset": 254,
        "startContainer": "./*[name()='p'][1]/text()[1]",
        "startOffset": 226,
        "type": "XpathRangeSelector",
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "Note@461.contents.keys",
            "observers": Set {},
          },
          "name": "Note@461.contents",
          "pendingKeys": Map {
            "selection" => false,
            "__jest-styled-components__" => false,
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            "content": "plants and animals appeared.",
            "endContainer": "./*[name()='p'][1]/text()[1]",
            "endOffset": 254,
            "startContainer": "./*[name()='p'][1]/text()[1]",
            "startOffset": 226,
            "type": "XpathRangeSelector",
            Symbol(mobx administration): [Circular],
          },
          "values": Map {
            "endOffset" => 254,
            "startOffset" => 226,
            "type" => "XpathRangeSelector",
            "content" => "plants and animals appeared.",
            "endContainer" => "./*[name()='p'][1]/text()[1]",
            "startContainer" => "./*[name()='p'][1]/text()[1]",
          },
        },
      },
      "courseId": undefined,
      "id": 3,
      "rect": Object {
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ObservableObject@463.keys",
            "observers": Set {},
          },
          "name": "ObservableObject@463",
          "pendingKeys": Map {
            "__jest-styled-components__" => false,
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            Symbol(mobx administration): [Circular],
          },
          "values": Map {},
        },
      },
      "section": undefined,
      "text": undefined,
      "title": undefined,
      "type": undefined,
      Symbol(mobx did run lazy initializers): true,
      Symbol(mobx administration): ObservableObjectAdministration$$1 {
        "defaultEnhancer": [Function],
        "keysAtom": Atom$$1 {
          "diffValue": 0,
          "isBeingObserved": false,
          "isPendingUnobservation": false,
          "lastAccessedBy": 0,
          "lowestObserverState": 2,
          "name": "Note@461.keys",
          "observers": Set {},
        },
        "name": "Note@461",
        "target": [Circular],
        "values": Map {
          "isNew" => false,
          "id" => 3,
          "chapter" => undefined,
          "courseId" => undefined,
          "section" => undefined,
          "title" => undefined,
          "text" => undefined,
          "type" => undefined,
          "rect" => Object {
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": -1,
                "name": "ObservableObject@463.keys",
                "observers": Set {},
              },
              "name": "ObservableObject@463",
              "pendingKeys": Map {
                "__jest-styled-components__" => false,
                "$$typeof" => false,
                "nodeType" => false,
                "@@__IMMUTABLE_ITERABLE__@@" => false,
                "@@__IMMUTABLE_RECORD__@@" => false,
                "_isMockFunction" => false,
                "toJSON" => false,
              },
              "proxy": [Circular],
              "target": Object {
                Symbol(mobx administration): [Circular],
              },
              "values": Map {},
            },
          },
          "contents" => Object {
            "content": "plants and animals appeared.",
            "endContainer": "./*[name()='p'][1]/text()[1]",
            "endOffset": 254,
            "startContainer": "./*[name()='p'][1]/text()[1]",
            "startOffset": 226,
            "type": "XpathRangeSelector",
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": 2,
                "name": "Note@461.contents.keys",
                "observers": Set {},
              },
              "name": "Note@461.contents",
              "pendingKeys": Map {
                "selection" => false,
                "__jest-styled-components__" => false,
                "$$typeof" => false,
                "nodeType" => false,
                "@@__IMMUTABLE_ITERABLE__@@" => false,
                "@@__IMMUTABLE_RECORD__@@" => false,
                "_isMockFunction" => false,
                "toJSON" => false,
              },
              "proxy": [Circular],
              "target": Object {
                "content": "plants and animals appeared.",
                "endContainer": "./*[name()='p'][1]/text()[1]",
                "endOffset": 254,
                "startContainer": "./*[name()='p'][1]/text()[1]",
                "startOffset": 226,
                "type": "XpathRangeSelector",
                Symbol(mobx administration): [Circular],
              },
              "values": Map {
                "endOffset" => 254,
                "startOffset" => 226,
                "type" => "XpathRangeSelector",
                "content" => "plants and animals appeared.",
                "endContainer" => "./*[name()='p'][1]/text()[1]",
                "startContainer" => "./*[name()='p'][1]/text()[1]",
              },
            },
          },
          "chapter_section" => ChapterSection {
            "chapter": 10,
            "section": 3,
            Symbol(mobx did run lazy initializers): true,
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": -1,
                "name": "ChapterSection@466.keys",
                "observers": Set {},
              },
              "name": "ChapterSection@466",
              "target": [Circular],
              "values": Map {
                "isNew" => false,
                "isEmpty" => false,
                "asString" => "10.3",
                "asArray" => Array [
                  10,
                  3,
                ],
              },
            },
          },
          "formattedChapterSection" => "undefined",
        },
      },
    },
  ],
  "2.2": Array [
    Note {
      "chapter": undefined,
      "chapter_section": ChapterSection {
        "chapter": 2,
        "section": 2,
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ChapterSection@446.keys",
            "observers": Set {},
          },
          "name": "ChapterSection@446",
          "target": [Circular],
          "values": Map {
            "isNew" => false,
            "isEmpty" => false,
            "asString" => "2.2",
            "asArray" => Array [
              2,
              2,
            ],
          },
        },
      },
      "contents": Object {
        "content": "plants and animals appeared.",
        "endContainer": "./*[name()='p'][1]/text()[1]",
        "endOffset": 254,
        "startContainer": "./*[name()='p'][1]/text()[1]",
        "startOffset": 226,
        "type": "XpathRangeSelector",
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "Note@441.contents.keys",
            "observers": Set {},
          },
          "name": "Note@441.contents",
          "pendingKeys": Map {
            "selection" => false,
            "__jest-styled-components__" => false,
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            "content": "plants and animals appeared.",
            "endContainer": "./*[name()='p'][1]/text()[1]",
            "endOffset": 254,
            "startContainer": "./*[name()='p'][1]/text()[1]",
            "startOffset": 226,
            "type": "XpathRangeSelector",
            Symbol(mobx administration): [Circular],
          },
          "values": Map {
            "endOffset" => 254,
            "startOffset" => 226,
            "type" => "XpathRangeSelector",
            "content" => "plants and animals appeared.",
            "endContainer" => "./*[name()='p'][1]/text()[1]",
            "startContainer" => "./*[name()='p'][1]/text()[1]",
          },
        },
      },
      "courseId": undefined,
      "id": 1,
      "rect": Object {
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ObservableObject@443.keys",
            "observers": Set {},
          },
          "name": "ObservableObject@443",
          "pendingKeys": Map {
            "__jest-styled-components__" => false,
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            Symbol(mobx administration): [Circular],
          },
          "values": Map {},
        },
      },
      "section": undefined,
      "text": undefined,
      "title": undefined,
      "type": undefined,
      Symbol(mobx did run lazy initializers): true,
      Symbol(mobx administration): ObservableObjectAdministration$$1 {
        "defaultEnhancer": [Function],
        "keysAtom": Atom$$1 {
          "diffValue": 0,
          "isBeingObserved": false,
          "isPendingUnobservation": false,
          "lastAccessedBy": 0,
          "lowestObserverState": 2,
          "name": "Note@441.keys",
          "observers": Set {},
        },
        "name": "Note@441",
        "target": [Circular],
        "values": Map {
          "isNew" => false,
          "id" => 1,
          "chapter" => undefined,
          "courseId" => undefined,
          "section" => undefined,
          "title" => undefined,
          "text" => undefined,
          "type" => undefined,
          "rect" => Object {
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": -1,
                "name": "ObservableObject@443.keys",
                "observers": Set {},
              },
              "name": "ObservableObject@443",
              "pendingKeys": Map {
                "__jest-styled-components__" => false,
                "$$typeof" => false,
                "nodeType" => false,
                "@@__IMMUTABLE_ITERABLE__@@" => false,
                "@@__IMMUTABLE_RECORD__@@" => false,
                "_isMockFunction" => false,
                "toJSON" => false,
              },
              "proxy": [Circular],
              "target": Object {
                Symbol(mobx administration): [Circular],
              },
              "values": Map {},
            },
          },
          "contents" => Object {
            "content": "plants and animals appeared.",
            "endContainer": "./*[name()='p'][1]/text()[1]",
            "endOffset": 254,
            "startContainer": "./*[name()='p'][1]/text()[1]",
            "startOffset": 226,
            "type": "XpathRangeSelector",
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": 2,
                "name": "Note@441.contents.keys",
                "observers": Set {},
              },
              "name": "Note@441.contents",
              "pendingKeys": Map {
                "selection" => false,
                "__jest-styled-components__" => false,
                "$$typeof" => false,
                "nodeType" => false,
                "@@__IMMUTABLE_ITERABLE__@@" => false,
                "@@__IMMUTABLE_RECORD__@@" => false,
                "_isMockFunction" => false,
                "toJSON" => false,
              },
              "proxy": [Circular],
              "target": Object {
                "content": "plants and animals appeared.",
                "endContainer": "./*[name()='p'][1]/text()[1]",
                "endOffset": 254,
                "startContainer": "./*[name()='p'][1]/text()[1]",
                "startOffset": 226,
                "type": "XpathRangeSelector",
                Symbol(mobx administration): [Circular],
              },
              "values": Map {
                "endOffset" => 254,
                "startOffset" => 226,
                "type" => "XpathRangeSelector",
                "content" => "plants and animals appeared.",
                "endContainer" => "./*[name()='p'][1]/text()[1]",
                "startContainer" => "./*[name()='p'][1]/text()[1]",
              },
            },
          },
          "chapter_section" => ChapterSection {
            "chapter": 2,
            "section": 2,
            Symbol(mobx did run lazy initializers): true,
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": -1,
                "name": "ChapterSection@446.keys",
                "observers": Set {},
              },
              "name": "ChapterSection@446",
              "target": [Circular],
              "values": Map {
                "isNew" => false,
                "isEmpty" => false,
                "asString" => "2.2",
                "asArray" => Array [
                  2,
                  2,
                ],
              },
            },
          },
          "formattedChapterSection" => "undefined",
        },
      },
    },
  ],
  "6.4": Array [
    Note {
      "chapter": undefined,
      "chapter_section": ChapterSection {
        "chapter": 6,
        "section": 4,
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ChapterSection@456.keys",
            "observers": Set {},
          },
          "name": "ChapterSection@456",
          "target": [Circular],
          "values": Map {
            "isNew" => false,
            "isEmpty" => false,
            "asString" => "6.4",
            "asArray" => Array [
              6,
              4,
            ],
          },
        },
      },
      "contents": Object {
        "content": "plants and animals appeared.",
        "endContainer": "./*[name()='p'][1]/text()[1]",
        "endOffset": 254,
        "startContainer": "./*[name()='p'][1]/text()[1]",
        "startOffset": 226,
        "type": "XpathRangeSelector",
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "Note@451.contents.keys",
            "observers": Set {},
          },
          "name": "Note@451.contents",
          "pendingKeys": Map {
            "selection" => false,
            "__jest-styled-components__" => false,
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            "content": "plants and animals appeared.",
            "endContainer": "./*[name()='p'][1]/text()[1]",
            "endOffset": 254,
            "startContainer": "./*[name()='p'][1]/text()[1]",
            "startOffset": 226,
            "type": "XpathRangeSelector",
            Symbol(mobx administration): [Circular],
          },
          "values": Map {
            "endOffset" => 254,
            "startOffset" => 226,
            "type" => "XpathRangeSelector",
            "content" => "plants and animals appeared.",
            "endContainer" => "./*[name()='p'][1]/text()[1]",
            "startContainer" => "./*[name()='p'][1]/text()[1]",
          },
        },
      },
      "courseId": undefined,
      "id": 2,
      "rect": Object {
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": -1,
            "name": "ObservableObject@453.keys",
            "observers": Set {},
          },
          "name": "ObservableObject@453",
          "pendingKeys": Map {
            "__jest-styled-components__" => false,
            "$$typeof" => false,
            "nodeType" => false,
            "@@__IMMUTABLE_ITERABLE__@@" => false,
            "@@__IMMUTABLE_RECORD__@@" => false,
            "_isMockFunction" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            Symbol(mobx administration): [Circular],
          },
          "values": Map {},
        },
      },
      "section": undefined,
      "text": undefined,
      "title": undefined,
      "type": undefined,
      Symbol(mobx did run lazy initializers): true,
      Symbol(mobx administration): ObservableObjectAdministration$$1 {
        "defaultEnhancer": [Function],
        "keysAtom": Atom$$1 {
          "diffValue": 0,
          "isBeingObserved": false,
          "isPendingUnobservation": false,
          "lastAccessedBy": 0,
          "lowestObserverState": 2,
          "name": "Note@451.keys",
          "observers": Set {},
        },
        "name": "Note@451",
        "target": [Circular],
        "values": Map {
          "isNew" => false,
          "id" => 2,
          "chapter" => undefined,
          "courseId" => undefined,
          "section" => undefined,
          "title" => undefined,
          "text" => undefined,
          "type" => undefined,
          "rect" => Object {
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": -1,
                "name": "ObservableObject@453.keys",
                "observers": Set {},
              },
              "name": "ObservableObject@453",
              "pendingKeys": Map {
                "__jest-styled-components__" => false,
                "$$typeof" => false,
                "nodeType" => false,
                "@@__IMMUTABLE_ITERABLE__@@" => false,
                "@@__IMMUTABLE_RECORD__@@" => false,
                "_isMockFunction" => false,
                "toJSON" => false,
              },
              "proxy": [Circular],
              "target": Object {
                Symbol(mobx administration): [Circular],
              },
              "values": Map {},
            },
          },
          "contents" => Object {
            "content": "plants and animals appeared.",
            "endContainer": "./*[name()='p'][1]/text()[1]",
            "endOffset": 254,
            "startContainer": "./*[name()='p'][1]/text()[1]",
            "startOffset": 226,
            "type": "XpathRangeSelector",
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": 2,
                "name": "Note@451.contents.keys",
                "observers": Set {},
              },
              "name": "Note@451.contents",
              "pendingKeys": Map {
                "selection" => false,
                "__jest-styled-components__" => false,
                "$$typeof" => false,
                "nodeType" => false,
                "@@__IMMUTABLE_ITERABLE__@@" => false,
                "@@__IMMUTABLE_RECORD__@@" => false,
                "_isMockFunction" => false,
                "toJSON" => false,
              },
              "proxy": [Circular],
              "target": Object {
                "content": "plants and animals appeared.",
                "endContainer": "./*[name()='p'][1]/text()[1]",
                "endOffset": 254,
                "startContainer": "./*[name()='p'][1]/text()[1]",
                "startOffset": 226,
                "type": "XpathRangeSelector",
                Symbol(mobx administration): [Circular],
              },
              "values": Map {
                "endOffset" => 254,
                "startOffset" => 226,
                "type" => "XpathRangeSelector",
                "content" => "plants and animals appeared.",
                "endContainer" => "./*[name()='p'][1]/text()[1]",
                "startContainer" => "./*[name()='p'][1]/text()[1]",
              },
            },
          },
          "chapter_section" => ChapterSection {
            "chapter": 6,
            "section": 4,
            Symbol(mobx did run lazy initializers): true,
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": -1,
                "name": "ChapterSection@456.keys",
                "observers": Set {},
              },
              "name": "ChapterSection@456",
              "target": [Circular],
              "values": Map {
                "isNew" => false,
                "isEmpty" => false,
                "asString" => "6.4",
                "asArray" => Array [
                  6,
                  4,
                ],
              },
            },
          },
          "formattedChapterSection" => "undefined",
        },
      },
    },
  ],
}
`;
